#### Please download all files in repository to same folder before running

In [1]:
# Import the custom vector module as vnp.. This will be used for calculating dot product in the end
import vector_numpy as vnp

In [2]:
import numpy as np
import string

In [3]:
# Read all 5 the text files
with open("The Tempest.txt") as fd1:
    tempest_text = fd1.read()
    

In [4]:
# Read all 5 the text files
with open("Romeo and Juliet.txt") as fd2:
    rj_text = fd2.read()

In [5]:
# Read all 5 the text files
with open("The Merchant of Venice.txt") as fd3:
    venice_text = fd3.read()

In [6]:
# Read all 5 the text files
with open("Hamlet.txt") as fd4:
    hamlet_text = fd4.read()

In [7]:
# Read all 5 the text files
with open("A Midsummer Night's Dream.txt") as fd5:
    midsum_text = fd5.read()

In [8]:
# Read file containing stopwords called english in same folder
# List of elizabethan stopwords taken from here: https://bryanbumgardner.com/elizabethan-stop-words-for-nlp/
# These new stop words have been updated in the stopwords file

with open("english.txt") as fd6:
    z1 = fd6.readlines()
    stop_words = [elem.strip() for elem in z1]

In [9]:
# Next steps will focus on breaking into words and cleaning the text

In [10]:
def clean_text(file_data):
    
    # Splitting the text by space
    words_temp_text = file_data
    words_temp_text1 = words_temp_text.split()
    
    # Change everything to lower case
    words_temp_text_lower = [elem.lower() for elem in words_temp_text1]
  
    # Strip
    words_temp_strip = [elem1.strip() for elem1 in words_temp_text_lower]
    
    # Drop last s (treat death and deaths as same... if julius becomes juliu,-  
    # - it doesnt matter because juliu will still be a relevant dimension across all
    # On testing with and without , dropping last s has improved scores, so retaining this logic
    words_temp_strip_s1 = [elemx for elemx in words_temp_strip if elemx[-1] != "s"]
    words_temp_strip_s2 = [elemy[:len(elemy)-1] for elemy in words_temp_strip if elemy[-1] == "s"]
    words_temp_strip_s1.extend(words_temp_strip_s1)
    
    # Remove stopwords
    words_temp_final = [elemz for elemz in words_temp_strip_s1 if elemz not in stop_words]
    
    # Remove punctuation
    table = str.maketrans('', '', string.punctuation)
    words_to_use = [w.translate(table) for w in words_temp_final]
    
    return words_to_use

In [11]:
cleaned_tempest = clean_text(tempest_text)
cleaned_rj = clean_text(rj_text)
cleaned_venice = clean_text(venice_text)
cleaned_hamlet = clean_text(hamlet_text)
cleaned_midsum = clean_text(midsum_text)

In [12]:
# Create a function to create a set of words
def set_words(ref_lst):
    ref_lst1 = ref_lst
    set1 = set()
    for elem2 in ref_lst1:
        if elem2 in set1:
            pass
        else:
            set1.add(elem2)
    return set1

In [13]:
set_tempest = set_words(cleaned_tempest)
set_rj = set_words(cleaned_rj)
set_hamlet = set_words(cleaned_hamlet)
set_venice = set_words(cleaned_venice)
set_midsum = set_words(cleaned_midsum)

In [14]:
union_all = set_tempest.union(set_rj, set_hamlet, set_venice, set_midsum)

In [15]:
#Convert the union set to list type
union_list = list(union_all)

In [16]:
# Function to make a vector, same dimensions 
def vecs(lst_to_vec):
    lst_to_vec1 = lst_to_vec
    main_vec = np.zeros(len(union_list), dtype = int)
    
    #IGNORE COMMENTED CODE
    #print(lst_to_vec1)
    #print(len(lst_to_vec1))
    #print(main_vec)
    #for i in range(len(union_list)):
    #    for j in range(len(lst_to_vec1)):
    #        if lst_to_vec1[j] == union_list[i]:
    #            main_vec[i] += 1
    #print(main_vec)
    #return main_vec

    lst_to_vec2 = np.array(lst_to_vec1)
    unique, counts = np.unique(lst_to_vec2, return_counts=True)
    dict1 = dict(zip(unique,counts))
    
    union_list_vec = np.array(union_list)
    
    for i in range(len(union_list_vec)):
        if union_list_vec[i] in dict1.keys():
            main_vec[i] = dict1[union_list_vec[i]]
    
    return main_vec


In [17]:
vec_tempest = vecs(cleaned_tempest)

In [18]:
vec_rj = vecs(cleaned_rj)

In [19]:
vec_hamlet = vecs(cleaned_hamlet)
vec_venice = vecs(cleaned_venice)

In [20]:
vec_midsum = vecs(cleaned_midsum)

#### To demonstrate the use of custom vector_numpy class/datatype, 
#### convert all previous 4 numpy arrays to lists.. and feed them to vector_numpy (vnp)

In [21]:
vec_tempest_list = list(vec_tempest)
vec_rj_list = list(vec_rj)
vec_hamlet_list = list(vec_hamlet)
vec_venice_list = list(vec_venice)
vec_midsum_list = list(vec_midsum)

In [22]:
# Pass lists to custom vector_numpy class

In [23]:
vec_tempest_custom = vnp.Vector(vec_tempest_list)
vec_rj_custom = vnp.Vector(vec_rj_list)
vec_hamlet_custom = vnp.Vector(vec_hamlet_list)
vec_venice_custom = vnp.Vector(vec_venice_list)
vec_midsum_custom = vnp.Vector(vec_midsum_list)

#### Tempest Versus rest
#### To demonstrate use of custom class, we will use it for dot product
#### However, we will still be using numpy directly to compute magnitude
#### Hence, while dot product, variable_custom has been used, as it is object of custom vector class
#### To compute magnitude using numpy, direct numpy arrays are used

In [24]:
print("Tempest versus RJ:")
print((vec_tempest_custom@vec_rj_custom)/((np.linalg.norm(vec_tempest))*(np.linalg.norm(vec_rj))))
print("Tempest versus Hamlet:")
print((vec_tempest_custom@vec_hamlet_custom)/((np.linalg.norm(vec_tempest))*(np.linalg.norm(vec_hamlet))))
print("Tempest versus Venice:")
print((vec_tempest_custom@vec_venice_custom)/((np.linalg.norm(vec_tempest))*(np.linalg.norm(vec_venice))))
print("Tempest versus Midsummer")
print((vec_tempest_custom@vec_midsum_custom)/((np.linalg.norm(vec_tempest))*(np.linalg.norm(vec_midsum))))
print("Tempest versus Tempest,it should be 1:")
print((vec_tempest_custom@vec_tempest_custom)/((np.linalg.norm(vec_tempest))*(np.linalg.norm(vec_tempest))))

Tempest versus RJ:
0.32995683904394574
Tempest versus Hamlet:
0.28675907778325527
Tempest versus Venice:
0.381014962676081
Tempest versus Midsummer
0.3617662920896585
Tempest versus Tempest,it should be 1:
1.0000000000000002


#### Romeo Juliet Versus rest

In [25]:
print("RJ versus Tempest:")
print((vec_tempest_custom@vec_rj_custom)/((np.linalg.norm(vec_tempest))*(np.linalg.norm(vec_rj))))
print("RJ versus Venice:")
print((vec_venice_custom@vec_rj_custom)/((np.linalg.norm(vec_venice))*(np.linalg.norm(vec_rj))))
print("RJ versus Hamlet:")
print((vec_hamlet_custom@vec_rj_custom)/((np.linalg.norm(vec_hamlet))*(np.linalg.norm(vec_rj))))
print("RJ versus Midsum")
print((vec_midsum_custom@vec_rj_custom)/((np.linalg.norm(vec_midsum))*(np.linalg.norm(vec_rj))))
print("RJ versus RJ, it should be 1:")
print((vec_rj_custom@vec_rj_custom)/((np.linalg.norm(vec_rj))*(np.linalg.norm(vec_rj))))

RJ versus Tempest:
0.32995683904394574
RJ versus Venice:
0.4332501084909256
RJ versus Hamlet:
0.2789928113917102
RJ versus Midsum
0.44789651850628714
RJ versus RJ, it should be 1:
0.9999999999999999


#### Venice Versus rest

In [26]:
print("Venice versus Hamlet:")
print((vec_venice_custom@vec_hamlet_custom)/((np.linalg.norm(vec_venice))*(np.linalg.norm(vec_hamlet))))
print("Venice versus RJ:")
print((vec_venice_custom@vec_rj_custom)/((np.linalg.norm(vec_venice))*(np.linalg.norm(vec_rj))))
print("Venice versus Tempest:")
print((vec_venice_custom@vec_tempest_custom)/((np.linalg.norm(vec_venice))*(np.linalg.norm(vec_tempest))))
print("Venice versus Midsum:")
print((vec_venice_custom@vec_midsum_custom)/((np.linalg.norm(vec_venice))*(np.linalg.norm(vec_midsum))))
print("Venice versus Venice: It should be 1")
print((vec_venice_custom@vec_venice_custom)/((np.linalg.norm(vec_venice))*(np.linalg.norm(vec_venice))))

Venice versus Hamlet:
0.33237605423779815
Venice versus RJ:
0.4332501084909256
Venice versus Tempest:
0.381014962676081
Venice versus Midsum:
0.4646048373068697
Venice versus Venice: It should be 1
1.0


#### Hamlet Versus Rest

In [27]:
print("Hamlet versus RJ:")
print((vec_hamlet_custom@vec_rj_custom)/((np.linalg.norm(vec_hamlet))*(np.linalg.norm(vec_rj))))
print("Hamlet versus Tempest:")
print((vec_hamlet_custom@vec_tempest_custom)/((np.linalg.norm(vec_hamlet))*(np.linalg.norm(vec_tempest))))
print("Hamlet versus Venice:")
print((vec_hamlet_custom@vec_venice_custom)/((np.linalg.norm(vec_hamlet))*(np.linalg.norm(vec_venice))))
print("Hamlet versus Midsum:")
print((vec_hamlet_custom@vec_midsum_custom)/((np.linalg.norm(vec_hamlet))*(np.linalg.norm(vec_midsum))))
print("Hamlet versus Hamlet,it should be 1:")
print((vec_hamlet_custom@vec_hamlet_custom)/((np.linalg.norm(vec_hamlet))*(np.linalg.norm(vec_hamlet))))

Hamlet versus RJ:
0.2789928113917102
Hamlet versus Tempest:
0.28675907778325527
Hamlet versus Venice:
0.33237605423779815
Hamlet versus Midsum:
0.3070852533049582
Hamlet versus Hamlet,it should be 1:
1.0


#### Midsummers Versus Rest

In [28]:
print("Midsummer versus Hamlet:")
print((vec_midsum_custom@vec_hamlet_custom)/((np.linalg.norm(vec_midsum))*(np.linalg.norm(vec_hamlet))))
print("Midsummer versus RJ:")
print((vec_midsum_custom@vec_rj_custom)/((np.linalg.norm(vec_midsum))*(np.linalg.norm(vec_rj))))
print("Midsummer versus Tempest:")
print((vec_midsum_custom@vec_tempest_custom)/((np.linalg.norm(vec_midsum))*(np.linalg.norm(vec_tempest))))
print("Midsummer versus Venice:")
print((vec_midsum_custom@vec_venice_custom)/((np.linalg.norm(vec_midsum))*(np.linalg.norm(vec_venice))))
print("Midsummer versus Midsummer,it should be 1:")
print((vec_midsum_custom@vec_midsum_custom)/((np.linalg.norm(vec_midsum))*(np.linalg.norm(vec_midsum))))

Midsummer versus Hamlet:
0.3070852533049582
Midsummer versus RJ:
0.44789651850628714
Midsummer versus Tempest:
0.3617662920896585
Midsummer versus Venice:
0.4646048373068697
Midsummer versus Midsummer,it should be 1:
1.0


## From Above Midsummers Night Dream and Merchan of Venice has max score of similarity ~ 0.46